--- 
title: "Salmonid immune study"
output: plot_heatmap_immmune.html
author: Joe Colgan (joscolgan)
---

## Introduction:  
This script generates a heatmap for significant differentially expressed genes identified as differentially expressed between male and female brown trout.
The script takes raw gene-level counts generated by [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html), scales counts, subsets rows by most significant genes, clusters based by expression similarity across genes and generates a heatmap.  
The output is a heatmap as .png, .jpeg and/or .pdf.

1. Load libraries:

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "reshape2",
               "DESeq2",
               "lintr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE)
    }
}
## Create output directory:
dir.create(path = "./results")
```

2. Examine count data:

```{r, message = FALSE}
## Normalise counts:
normalised_counts <- read.table(file = "input/normalised_counts.txt",
                                header = TRUE)
## Convert to dataframe:
normalised_counts_df <- as.data.frame(normalised_counts)
## Reorder based on sex:
male_samples <- c(1,2,3,4,9,10,11,12,13,15:17,26,28,32)
female_samples <- c(5:8,14,18:25,27,29:31,33:37)
## Reorder the columns by phenotype:
normalised_counts_reordered <- normalised_counts_df[, c(male_samples,
                                                        female_samples)]
```

3. Cluster samples by similarity in gene expression profiles:

```{r, message = FALSE}
immune_genes_df <- read.table(file = "data/immune_genes_plus_pvalue_new.txt",
                              header = FALSE,
                              col.names = c("locus",
                                            "pvalue"))
immune_genes_sig <- subset(immune_genes_df,
                           pvalue < 0.01)
```

4. Extract normalised counts for genes of interest:

```{r, message = FALSE}
## Subset rows containing gene names of interest:
immune_genes_sig_counts <- normalised_counts_reordered[match(immune_genes_sig$locus,
                                                              row.names(normalised_counts_reordered)), ]
immune_genes_sig_counts$gene_id <- row.names(immune_genes_sig_counts)

## Update sample name for plotting:
for (name in grep("_M_", colnames(immune_genes_sig_counts))) {
        print(name)
        colnames(immune_genes_sig_counts)[[name]] <- paste("Male ",
                                                           name,
                                                           sep = "")
}

for (name in grep("_F_", colnames(immune_genes_sig_counts))) {
        print(name)
        mod_name <- name - 15
        colnames(immune_genes_sig_counts)[[name]] <- paste("Female ",
                                                           mod_name,
                                                           sep = "")
}

immune_genes_sig_counts$gene_id <- NULL
## Cluster based on euclidean distance:
## Reorder:
dat <- immune_genes_sig_counts
row_order <- hclust(dist(dat))$order # clustering
col_order <- hclust(dist(t(immune_genes_sig_counts)))$order
dat_new <- immune_genes_sig_counts[row_order, ] # re-order matrix accoring to clustering
dat_new_df <- as.data.frame(dat_new)
dat_new_df$samples <- rownames(dat_new_df)

dat_new_melted <- melt(dat_new_df,
                       id.vars = "samples")

dat_new_melted_ordered <- dat_new_melted[order(dat_new_melted$variable), ]

## Rename:
colnames(dat_new_melted_ordered) <- c("samples",
                                      "variable",
                                      "value")
## Adjust levels:
dat_new_melted_ordered$samples <- factor(dat_new_melted_ordered$samples,
                                 levels = c(unique(as.character(unlist(dat_new_melted_ordered$samples)))))
dat_new_melted_ordered$variable <- factor(as.character(dat_new_melted_ordered$variable),
                                          levels = c(unique(as.character(unlist(dat_new_melted_ordered$variable)))))

dat_new_melted_ordered$value <- as.numeric(unlist(dat_new_melted_ordered$value))

heatmap_plot <- ggplot(dat_new_melted_ordered,
                       aes(variable, samples)) +
        geom_tile(aes(fill = value),
                  colour = "white") +
        geom_vline(xintercept = c(0.5,
                                  15.5,
                                  37.5)) +
        scale_fill_gradient2(low = "black",
                             mid = "white",
                             high = "red") +
        theme(legend.title = element_text(face = "bold",
                                          size = 12),
              legend.text = element_text(size = 12),
              axis.title = element_blank(),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank(),
              axis.text.x = element_text(face = "bold",
                                         size = 12,
                                         angle = 45,
                                         hjust = 1)) +
        theme(legend.position = "top")

heatmap_plot <- heatmap_plot +
        labs(fill = "Normalised gene-level counts")

## Save image:
output  <- "results"
ggsave(file = paste(output,
                    "/",
                    "heatmap_male_female_immune.pdf",
                    sep = ""),
       dpi = 600,
       height = 8,
       width = 12)

ggsave(file = paste(output,
                    "/",
                    "heatmap_male_female_immune.tiff",
                    sep = ""),
       height = 8,
       width = 12)

ggsave(file = paste(output,
                    "/",
                    "heatmap_male_female_immune.jpeg",
                    sep = ""),
       dpi = 600,
       height = 8,
       width = 12)

output  <- "results"
save.image(file = paste(output,
                    "/",
                    "heatmap_male_female_immune.RData",
                    sep = ""))
```
