
1. Load libraries:

```{r, message = FALSE}
require(ggplot2)
require(biomaRt)
```

2. Load input data:

```{r, message = FALSE}
## Load R object:
load(file = "results/putative_strutta_immune_genes.RData")
```

3. Extract paralogues for brown trout:

```{r, message = FALSE}
## Use the latest ensembl biomart database:
ensembl <- useMart("ensembl")

## Extract information related to brown trout (salmo trutta):
ensembl <- useDataset("strutta_gene_ensembl",
                      mart = ensembl)

## Extract paralog
attrib <- listAttributes(ensembl,
                         page = "homologs")

hgid <- getBM(attributes = "ensembl_gene_id",
              filters    = "with_strutta_paralog",
              values     = TRUE,
              mart       = ensembl)$ensembl_gene_id

para_comp <- getBM(attributes = c('ensembl_gene_id',
                                  'chromosome_name',
                                  'start_position',
                                  'end_position',
                                  'strutta_paralog_ensembl_gene',
                                  'strutta_paralog_associated_gene_name',
                                  'strutta_paralog_chromosome',
                                  'strutta_paralog_chrom_start',
                                  'strutta_paralog_chrom_end',
                                  'strutta_paralog_subtype',
                                  'strutta_paralog_orthology_type',
                                  'strutta_paralog_perc_id',
                                  'strutta_paralog_perc_id_r1'),
                   filters = 'ensembl_gene_id',
                   values = unique_gene_list,
                   mart = ensembl)

## Subset within species paralog:
para_comp_within_species <- subset(para_comp,
                                   strutta_paralog_orthology_type == "within_species_paralog")
```

4. There are different subtypes of paralogs that can be subset:
- First subtype allows for identifying species-specific paralogs:

```{r, message = FALSE}
para_comp_within_strutta <- subset(para_comp_within_species,
                                   strutta_paralog_subtype == "Salmo trutta")

## Calculate the average percentage identify between matches:
para_comp_within_strutta$average_perc_id <- (para_comp_within_strutta$strutta_paralog_perc_id +
                                             para_comp_within_strutta$strutta_paralog_perc_id_r1) / 2

## Order based on average percentage identity:
para_comp_within_strutta_ordered <- para_comp_within_strutta[order(-para_comp_within_strutta$average_perc_id), ]

## Take unique matches:
para_comp_within_strutta_match <- para_comp_within_strutta_ordered[match(unique(para_comp_within_strutta_ordered$ensembl_gene_id),
                                   para_comp_within_strutta_ordered$ensembl_gene_id), ]

## Calculate gene size for each putative ohnolog:
## For gene 1: Currently do not evolutionary distance inforation so name is arbitrary:
para_comp_within_strutta_match$gene_1_size <- para_comp_within_strutta_match$end_position - para_comp_within_strutta_match$start_position
## For gene 2: Currently do not evolutionary distance inforation so name is arbitrary:
para_comp_within_strutta_match$gene_2_size <- para_comp_within_strutta_match$strutta_paralog_chrom_end - para_comp_within_strutta_match$strutta_paralog_chrom_start

## Subset matches with high sequence similarity:
para_comp_within_strutta_match_high_sim <- subset(para_comp_within_strutta_match,
                                                     average_perc_id >= 85)

## Check how many overlap with putative immune genes:
nrow(subset(para_comp_within_strutta_match_high_sim,
       ensembl_gene_id %in% unique_gene_list))

nrow(subset(para_comp_within_strutta_match_high_sim,
       strutta_paralog_ensembl_gene %in% unique_gene_list))

## Look for overlap:
length(intersect(row.names(subset(para_comp_within_strutta_match_high_sim,
                           ensembl_gene_id %in% unique_gene_list)),
          row.names(subset(para_comp_within_strutta_match_high_sim,
                           strutta_paralog_ensembl_gene %in% unique_gene_list))))

## Write to file:
write.table(x = para_comp_within_strutta_match_high_sim,
            file = "results/strutta_immune_duplicates.txt",
            col.names = TRUE,
            row.names = FALSE,
            quote = FALSE,
            sep = "\t")
```

5. There are different subtypes of paralogs that can be subset:
- Second subtype allows for identifying genus-specific paralogs:

```{r, message = TRUE}
para_comp_within_salmo <- subset(para_comp_within_species,
          strutta_paralog_subtype == "Salmo")

## Calculate the average percentage identify between matches:
para_comp_within_salmo$average_perc_id <- (para_comp_within_salmo$strutta_paralog_perc_id +
                                           para_comp_within_salmo$strutta_paralog_perc_id_r1) / 2

## Order based on average percentage identity:
para_comp_within_salmo_ordered <- para_comp_within_salmo[order(-para_comp_within_salmo$average_perc_id), ]

## Take unique matches:
para_comp_within_salmo_match <- para_comp_within_salmo_ordered[match(unique(para_comp_within_salmo_ordered$ensembl_gene_id),
                                       para_comp_within_salmo_ordered$ensembl_gene_id), ]

## Check if on the same chromosome:
table(para_comp_within_salmo_match$chromosome_name == para_comp_within_salmo_match$strutta_paralog_chromosome)

## Calculate gene size:
para_comp_within_salmo_match$gene_1_size <- para_comp_within_salmo_match$end_position - para_comp_within_salmo_match$start_position
para_comp_within_salmo_match$gene_2_size <- para_comp_within_salmo_match$strutta_paralog_chrom_end - para_comp_within_salmo_match$strutta_paralog_chrom_start

## Calculate difference in gene sizes:
para_comp_within_salmo_match$gene_size <- abs(para_comp_within_salmo_match$gene_1_size - para_comp_within_salmo_match$gene_2_size)

## Subset matches with high sequence similarity:
para_comp_within_salmo_match_high_sim <- subset(para_comp_within_salmo_match,
                                                     average_perc_id >= 85)

## Check how many overlap with putative immune genes:
nrow(subset(para_comp_within_salmo_match_high_sim,
       ensembl_gene_id %in% unique_gene_list))

nrow(subset(para_comp_within_salmo_match_high_sim,
       strutta_paralog_ensembl_gene %in% unique_gene_list))

## Look for overlap:
length(intersect(row.names(subset(para_comp_within_salmo_match_high_sim,
                           ensembl_gene_id %in% unique_gene_list)),
          row.names(subset(para_comp_within_salmo_match_high_sim,
                           strutta_paralog_ensembl_gene %in% unique_gene_list))))

## Write to file:
write.table(x = para_comp_within_salmo_match_high_sim,
            file = "results/salmo_immune_duplicates.txt",
            col.names = TRUE,
            row.names = FALSE,
            quote = FALSE,
            sep = "\t")
```

6. There are different subtypes of paralogs that can be subset:
- Third subtype allows for identifying family-specific paralogs:

```{r, message = FALSE}
options(scipen = 666)
para_comp_within_salmoninae <- subset(para_comp_within_species,
          strutta_paralog_subtype == "Salmoninae")

## Calculate the average percentage identify between matches:
para_comp_within_salmoninae$average_perc_id <- (para_comp_within_salmoninae$strutta_paralog_perc_id +
                                                para_comp_within_salmoninae$strutta_paralog_perc_id_r1) / 2

## Order based on average percentage identity:
para_comp_within_salmoninae_ordered <- para_comp_within_salmoninae[order(-para_comp_within_salmoninae$average_perc_id), ]

## Take unique matches:
para_comp_within_salmoninae_match <- para_comp_within_salmoninae_ordered[match(unique(para_comp_within_salmoninae_ordered$ensembl_gene_id),
                                       para_comp_within_salmoninae_ordered$ensembl_gene_id), ]

## Calculate gene size:
para_comp_within_salmoninae_match$gene_1_size <- para_comp_within_salmoninae_match$end_position - para_comp_within_salmoninae_match$start_position
para_comp_within_salmoninae_match$gene_2_size <- para_comp_within_salmoninae_match$strutta_paralog_chrom_end - para_comp_within_salmoninae_match$strutta_paralog_chrom_start

## Calculate difference in gene sizes:
para_comp_within_salmoninae_match$gene_size <- abs(para_comp_within_salmoninae_match$gene_1_size - para_comp_within_salmoninae_match$gene_2_size)

## High sequence similarity:
para_comp_within_salmoninae_match_high_sim <- subset(para_comp_within_salmoninae_match,
                                                     average_perc_id >= 85)

## Plot:
ggplot(data = para_comp_within_salmoninae_match,
       aes(x = gene_size,
           y = average_perc_id)) +
        geom_point() +
        geom_point(data = para_comp_within_salmoninae_match_high_sim,
                   aes(x = gene_size,
                       y = average_perc_id),
                   colour = "red")

## Average difference in gene size:
summary(para_comp_within_salmoninae_match_high_sim$gene_size)
```

7. This last group in particular we want to compare with orthologs in pike:
- Here we are looking for high confident orthologs where there are two matches in the brown trout genome:

```{r, message = FALSE}
## Identify ensembl gene ids:
pike_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'elucius_homolog_ensembl_gene',
                     'elucius_homolog_chromosome',
                     'elucius_homolog_chrom_start',
                     'elucius_homolog_chrom_end',
                     'elucius_homolog_orthology_type',
                     'elucius_homolog_perc_id',
                     'elucius_homolog_perc_id_r1',
                     'elucius_homolog_goc_score',
                     #'mmusculus_homolog_wga_coverage',
                     'elucius_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
pike_comp_sub <- subset(pike_comp,
                        elucius_homolog_orthology_confidence == "1")

## Many to one:
pike_comp_sub_one2many <- subset(pike_comp_sub,
                                 elucius_homolog_orthology_type == "ortholog_one2many")

## Calculate average percentage similarity:
pike_comp_sub_one2many$average_perc_id <- (pike_comp_sub_one2many$elucius_homolog_perc_id + pike_comp_sub_one2many$elucius_homolog_perc_id_r1) / 2

## Subset putative immune genes:
pike_comp_sub_one2many_immune <- subset(pike_comp_sub_one2many,
                                        ensembl_gene_id %in% unique_gene_list &
                                                average_perc_id >= 75)
               
## Calculate the frequency of pike genes:
pike_comp_sub_one2many_immune_freq <- as.data.frame(table(sort(pike_comp_sub_one2many_immune$elucius_homolog_ensembl_gene)))

## Subset duplicated entries for pike:
pike_comp_sub_one2two_immune <- subset(pike_comp_sub_one2many_immune_freq,
                                       Freq == 2)$Var1

## Generate a list of putative immune genes duplicated in brown trout compared to pike: 
pike_comp_sub_one2two_immune_list <- as.character(unlist(pike_comp_sub_one2two_immune))

## Subset salmonid genes:
pike_comp_sub_one2many_immune_ohno <- subset(pike_comp_sub_one2many_immune,
       elucius_homolog_ensembl_gene %in% pike_comp_sub_one2two_immune_list)

## Order by pike gene:
pike_comp_sub_one2many_immune_ohno <- pike_comp_sub_one2many_immune_ohno[order(pike_comp_sub_one2many_immune_ohno$elucius_homolog_ensembl_gene), ]
```

8. Calculate gene size for salmo trutta:

```{r, message = FALSE}
## Calculate gene size:
pike_comp_sub_one2many_immune_ohno$strutta_gene_size <- pike_comp_sub_one2many_immune_ohno$end_position - pike_comp_sub_one2many_immune_ohno$start_position
```

9. Write to file:

```{r, message = FALSE}
write.table(pike_comp_sub_one2many_immune_ohno,
            file = "results/pike_comp_sub_one2many_immune_ohno_strutta.txt",
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")

## Save as an object:
save.image(file = "results/pike_comp_sub_one2many_immune_ohno_strutta.RData")
```

10. Check style guide:

```{r, message = FALSE}
lintr::lint(file = "identify_immune_paralogs_ensembl_strutta.Rmd")
```