


1. Load libraries:

```{r, message = FALSE}
require(ggplot2)
require(ggpubr)
require(dplyr)
```

2. Load and subset:

```{r, message = FALSE}
diverged_copy_coordinates <- read.table(file = "input/diverged_gene_coordinates_plus_genes.txt",
                                        header = FALSE,
                                        col.names = c("diverged_locus",
                                                      "diverged_chromosome",
                                                     "diverged_start",
                                                     "diverged_end"))

## Rename locus:
diverged_copy_coordinates$diverged_locus <- gsub(pattern = "[.]1",
                                                 replacement = "",
                                                 diverged_copy_coordinates$diverged_locus)

conserved_copy_coordinates <- read.table(file = "input/conserved_gene_coordinates_plus_genes.txt",
                                         header = FALSE,
                                        col.names = c("conserved_locus",
                                                      "conserved_chromosome",
                                                     "conserved_start",
                                                     "conserved_end"))

## Rename locus:
conserved_copy_coordinates$conserved_locus <- gsub(pattern = "[.]1",
                                                 replacement = "",
                                                 conserved_copy_coordinates$conserved_locus)

## Check synteny:
combined_df <- cbind(diverged_copy_coordinates,
                     conserved_copy_coordinates)

combined_ordered <- combined_df[
  with(combined_df, order(diverged_chromosome,
                          diverged_start)),
]

combined_ordered <- subset(combined_ordered,
                            as.numeric(diverged_chromosome) <= 40 &
                            as.numeric(conserved_chromosome) <= 40)

combined_ordered$diverged_chromosome <- factor(combined_ordered$diverged_chromosome,
                               levels = unique(as.numeric(combined_ordered$diverged_chromosome)))

combined_ordered$conserved_chromosome <- factor(combined_ordered$conserved_chromosome,
                               levels = sort(unique(as.numeric(combined_ordered$diverged_chromosome))))

ohnolog_scatterplot <- ggplot(data = combined_ordered,
       aes(x = conserved_chromosome,
           y = diverged_chromosome)) +
        geom_point() +
        xlab(label = "Chromosome with diverged ohnolog copy") +
        ylab(label = "Chromosome with conserved ohnolog copy") +
        theme_bw() +
        theme(axis.title = element_text(face = "bold",
                                        size = 12),
              axis.text = element_text(face = "plain",
                                       size = 10))
```

3. Load differential expression analysis results:

```{r, message = FALSE}
## Read in DEG information:
deg_information <- read.table(file = "data/differentially_expressed_genes.txt",
                              header = TRUE)
row.names(deg_information) <- gsub(pattern = "[.]1",
                                   replacement = "",
                                   row.names(deg_information))

## Gene positions:
gene_location_info <- read.table(file = "data/gene_positions.txt",
                                 header = FALSE,
                                 col.names = c("chrom",
                                               "start",
                                               "end",
                                               "locus"))

## Convert "chrom" to numeric value:
gene_location_info$chrom <- as.numeric(unlist(gene_location_info$chrom))


deg_matched <- gene_location_info[match(row.names(deg_information),
                                gene_location_info$locus), ]

deg_matched$log2FC <- deg_information$log2FoldChange

deg_matched <- subset(deg_matched,
                      chrom <= 40)

## columns required for Manhattan plot:
input_for_manhattan <- deg_matched %>% select(chrom,
                                              locus,
                                              start,
                                              log2FC)

## Filter based on chromosome:
combined_filtered <- subset(input_for_manhattan,
                            as.numeric(chrom) <= 40)

## Sort by chromosome and position:
combined_sorted <- combined_filtered[
  with(combined_filtered, order(chrom,
                                start)),
]

## Convert CHR to factor and sort levels by correct order:
combined_sorted$chrom <- factor(combined_sorted$chrom,
                               levels = unique(as.numeric(combined_sorted$chrom)))
options(scipen = 999)

don <- combined_sorted %>%
  # Compute chromosome size
  group_by(chrom) %>%
  summarise(chr_len = max(start)) %>%
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(as.numeric(chr_len)) - chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(combined_sorted, ., by = c("chrom" = "chrom")) %>%
  # Add a cumulative position of each SNP
  arrange(chrom, start) %>%
  mutate(BPcum = start + tot) %>%
  # Add highlight and annotation information
  mutate(is_diverged = ifelse(locus %in% diverged_copy_coordinates$diverged_locus,
                                "yes",
                                "no")) %>%
  mutate(is_conserved = ifelse(locus %in% conserved_copy_coordinates$conserved_locus,
                                "yes",
                                "no"))

axisdf <- don %>% group_by(chrom) %>%
          summarize(center = (max(BPcum) + min(BPcum)) / 2)


# Make the plot
con_manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = abs(log2FC))) +
    # Show all points
    geom_point(aes(color = as.factor(chrom)),
                alpha = 0.3,
                size = .3) +
    scale_color_manual(values = rep(c("grey",
                                      "black"),
                                    22)) +
    xlab("Chromosome") +
    ylab("Log2FoldChange") +
    # custom X axis:
    scale_x_continuous(label = axisdf$chrom,
                       breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    # Add highlighted points
    geom_point(data = subset(don,
                             is_conserved == "yes"),
               color = "darkorange",
               alpha = 0.8,
               size = 1.3) +
    # Custom the theme:
    theme_bw() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text.x = element_blank(),
      axis.text.y = element_text(face = "plain",
                                  size = 10),
      axis.ticks.x = element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_text(face = "bold",
                                  size = 12))

# Make the plot
div_manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = abs(log2FC))) +
    # Show all points
    geom_point(aes(color = as.factor(chrom)),
                alpha = 0.3,
                size = .3) +
    scale_color_manual(values = rep(c("grey",
                                      "black"),
                                    22)) +
    xlab("Chromosome") +
    ylab("Log2FoldChange") +
    # custom X axis:
    scale_x_continuous(label = axisdf$chrom,
                       breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    # Add highlighted points
    geom_point(data = subset(don,
                             is_diverged == "yes"),
               color = "red",
               alpha = 0.8,
               size = 1.3) +
    # Custom the theme:
    theme_bw() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text = element_text(size = 10,
                               face = "plain"),
      axis.title = element_text(size = 15,
                              face = "bold"))
```

4. Generate a combined plot:

```{r, message = FALSE}
ggarrange(ohnolog_scatterplot,
          con_manhattan_plot,
       div_manhattan_plot,
       nrow = 3,
       ncol = 1,
       labels = c("A",
                  "B",
                  "C"),
       align = "hv")
```


