

```{r, message = FALSE}
require(biomaRt)
require(ggplot2)
listMarts()

ensembl <- useMart("ensembl")

datasets <- listDatasets(ensembl)

ensembl = useDataset("ssalar_gene_ensembl",
                     mart = ensembl)
```

```{r, message = FALSE}
## Identify NCBI gene ids:
salmo_ncbi <- getBM(attributes = c('ensembl_gene_id',
                                   'entrezgene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position'), 
      mart = ensembl)
```


Subset salmon orthologs:

```{r, message = FALSE}
salmo_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'strutta_homolog_ensembl_gene',
                     'strutta_homolog_chromosome',
                     'strutta_homolog_chrom_start',
                     'strutta_homolog_chrom_end',
                     'strutta_homolog_orthology_type',
                     'strutta_homolog_perc_id',
                     'strutta_homolog_perc_id_r1',
                     'strutta_homolog_goc_score',
                     'strutta_homolog_wga_coverage',
                     'strutta_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
salmo_comp_sub <- subset(salmo_comp,
       strutta_homolog_orthology_confidence == "1")

## Calculate the average percentage between matches:
salmo_comp_sub$average_perc_id <- (salmo_comp_sub$strutta_homolog_perc_id + salmo_comp_sub$strutta_homolog_perc_id_r1) / 2
summary(salmo_comp_sub$average_perc_id)

require(ggplot2)
## Plot percentage similarity against each other and highlight "confident" calls:
ggplot(data = salmo_comp,
       aes(x = strutta_homolog_perc_id_r1,
           y = strutta_homolog_perc_id)) +
        geom_point() +
        geom_point(data = salmo_comp_sub,
                   aes(x = strutta_homolog_perc_id_r1,
                       y = strutta_homolog_perc_id),
                   colour = "blue")
```

Subset Danio rerio genes:

```{r, message = FALSE}
## For putative mouse immune genes, gene ids were taken from MIG:
zfish_immune_genes <- scan(file = "data/danio_rerio/drerio_immune_genes.txt",
                           as.character())

## Identify ensembl gene ids:
zfish_ensembl = useDataset("drerio_gene_ensembl",
                     mart = ensembl)

## Extract mig ids for each ensembl annotated gene:
zfin_to_ensembl <- getBM(attributes = c('ensembl_gene_id',
                     "zfin_id_id"),
                     mart = zfish_ensembl)

zfin_to_ensembl_immune <- subset(zfin_to_ensembl,
                                zfin_id_id %in% zfish_immune_genes)

danio_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'drerio_homolog_ensembl_gene',
                     'drerio_homolog_chromosome',
                     'drerio_homolog_chrom_start',
                     'drerio_homolog_chrom_end',
                     'drerio_homolog_orthology_type',
                     'drerio_homolog_perc_id',
                     'drerio_homolog_perc_id_r1',
                     'drerio_homolog_goc_score',
                     'drerio_homolog_wga_coverage',
                     'drerio_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
danio_comp_sub <- subset(danio_comp,
       drerio_homolog_orthology_confidence == "1")

## Calculate the average percentage between matches:
danio_comp_sub$average_perc_id <- (danio_comp_sub$drerio_homolog_perc_id + danio_comp_sub$drerio_homolog_perc_id_r1) / 2
summary(danio_comp_sub$average_perc_id)

require(ggplot2)
## Plot percentage similarity against each other and highlight "confident" calls:
ggplot(data = danio_comp,
       aes(x = drerio_homolog_perc_id_r1,
           y = drerio_homolog_perc_id)) +
        geom_point() +
        geom_point(data = danio_comp_sub,
                   aes(x = drerio_homolog_perc_id_r1,
                       y = drerio_homolog_perc_id),
                   colour = "blue")

## Subset putative immmune homologs:
danio_comp_immune <- subset(danio_comp_sub,
                            drerio_homolog_ensembl_gene %in% zfin_to_ensembl_immune$ensembl_gene_id)

summary(danio_comp_immune$average_perc_id)
```

Subset Mus musculus genes:

```{r, message = FALSE}
## For putative mouse immune genes, gene ids were taken from MIG:
mouse_immune_genes <- scan(file = "data/mus_musculus/mig_immune_genes.tmp",
                           as.character())

## Identify ensembl gene ids:
mus_ensembl = useDataset("mmusculus_gene_ensembl",
                     mart = ensembl)

## Extract mig ids for each ensembl annotated gene:
mig_to_ensembl <- getBM(attributes = c('ensembl_gene_id',
                     "mgi_id"),
                     mart = mus_ensembl)


mig_to_ensembl_immune <- subset(mig_to_ensembl,
                                mgi_id %in% mouse_immune_genes)

mus_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'mmusculus_homolog_ensembl_gene',
                     'mmusculus_homolog_chromosome',
                     'mmusculus_homolog_chrom_start',
                     'mmusculus_homolog_chrom_end',
                     'mmusculus_homolog_orthology_type',
                     'mmusculus_homolog_perc_id',
                     'mmusculus_homolog_perc_id_r1',
                     'mmusculus_homolog_goc_score',
                     #'mmusculus_homolog_wga_coverage',
                     'mmusculus_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
mus_comp_sub <- subset(mus_comp,
       mmusculus_homolog_orthology_confidence == "1")

## Calculate the average percentage between matches:
mus_comp_sub$average_perc_id <- (mus_comp_sub$mmusculus_homolog_perc_id + mus_comp_sub$mmusculus_homolog_perc_id_r1) / 2
summary(mus_comp_sub$average_perc_id)

require(ggplot2)
## Plot percentage similarity against each other and highlight "confident" calls:
ggplot(data = mus_comp,
       aes(x = mmusculus_homolog_perc_id_r1,
           y = mmusculus_homolog_perc_id)) +
        geom_point() +
        geom_point(data = mus_comp_sub,
                   aes(x = mmusculus_homolog_perc_id_r1,
                       y = mmusculus_homolog_perc_id),
                   colour = "blue")

## Subset putative immmune homologs:
mus_comp_immune <- subset(mus_comp_sub,
                          mmusculus_homolog_ensembl_gene %in% mig_to_ensembl_immune$ensembl_gene_id)

summary(mus_comp_immune$average_perc_id)
```

Subset human genes:

```{r, message = FALSE}
## For putative mouse immune genes, gene ids were taken from MIG:
human_immune_genes <- scan(file = "data/homo_sapiens/human_immune_gene_ids.txt",
                           as.character())

## Identify ensembl gene ids:
homo_ensembl = useDataset("hsapiens_gene_ensembl",
                     mart = ensembl)

## Extract mig ids for each ensembl annotated gene:
hgnc_to_ensembl <- getBM(attributes = c('ensembl_gene_id',
                     "hgnc_symbol"),
                     mart = homo_ensembl)

hgnc_to_ensembl_immune <- subset(hgnc_to_ensembl,
                                hgnc_symbol %in% human_immune_genes)

human_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'hsapiens_homolog_ensembl_gene',
                     'hsapiens_homolog_chromosome',
                     'hsapiens_homolog_chrom_start',
                     'hsapiens_homolog_chrom_end',
                     'hsapiens_homolog_orthology_type',
                     'hsapiens_homolog_perc_id',
                     'hsapiens_homolog_perc_id_r1',
                     'hsapiens_homolog_goc_score',
                     #'mmusculus_homolog_wga_coverage',
                     'hsapiens_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
human_comp_sub <- subset(human_comp,
       hsapiens_homolog_orthology_confidence == "1")

## Calculate the average percentage between matches:
human_comp_sub$average_perc_id <- (human_comp_sub$hsapiens_homolog_perc_id + human_comp_sub$hsapiens_homolog_perc_id_r1) / 2
summary(human_comp_sub$average_perc_id)

require(ggplot2)
## Plot percentage similarity against each other and highlight "confident" calls:
ggplot(data = human_comp,
       aes(x = hsapiens_homolog_perc_id_r1,
           y = hsapiens_homolog_perc_id)) +
        geom_point() +
        geom_point(data = human_comp_sub,
                   aes(x = hsapiens_homolog_perc_id_r1,
                       y = hsapiens_homolog_perc_id),
                   colour = "blue")

## Subset putative immmune homologs:
human_comp_immune <- subset(human_comp_sub,
                          hsapiens_homolog_ensembl_gene %in% hgnc_to_ensembl_immune$ensembl_gene_id)

summary(human_comp_immune$average_perc_id)
```

Check overlap:

```{r, message = FALSE}
require(eulerr)

danio_comp_immune_id <- as.character(danio_comp_immune$ensembl_gene_id)
mus_comp_immune_id <- as.character(mus_comp_immune$ensembl_gene_id)
homo_comp_immune_id <- as.character(human_comp_immune$ensembl_gene_id)

venndiagram <- euler(combinations = list("drerio_strutta" = unique(danio_comp_immune_id),
                                        "mmusculus_strutta" = unique(mus_comp_immune_id),
                                        "hsapiens_strutta" = unique(homo_comp_immune_id)))

plot(venndiagram,
     quantities = TRUE,
     edges = TRUE)
```

Output nonredundant list:

```{r, message = FALSE}
unique_gene_list <- unique(sort(c(danio_comp_immune_id,
         mus_comp_immune_id,
         homo_comp_immune_id)))

length(unique_gene_list)

## Write to file:
write(unique_gene_list,
      file = "results/putative_ssalar_immune_genes.txt")
```

Check paralogues for salmon:

```{r, message = FALSE}
attrib <- listAttributes(ensembl, page = "homologs")

hgid <- getBM(attributes = "ensembl_gene_id",
              filters    = "with_ssalar_paralog",
              values     = TRUE,
              mart       = ensembl)$ensembl_gene_id

para_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'ssalar_paralog_ensembl_gene',
                     'ssalar_paralog_associated_gene_name',
                     'ssalar_paralog_chromosome',
                     'ssalar_paralog_chrom_start',
                     'ssalar_paralog_chrom_end',
                     'ssalar_paralog_subtype',
                     'ssalar_paralog_orthology_type',
                     'ssalar_paralog_perc_id',
                     #'mmusculus_homolog_wga_coverage',
                     'ssalar_paralog_perc_id_r1'), 
                   filters = 'ensembl_gene_id',
                   values = unique_gene_list,
      mart = ensembl)

## Subset within species paralog:
para_comp_within_species <- subset(para_comp,
       ssalar_paralog_orthology_type == "within_species_paralog")

## 
para_comp_within_ssalar <- subset(para_comp_within_species,
          ssalar_paralog_subtype == "Salmo trutta")
para_comp_within_ssalar$average_perc_id <- (para_comp_within_ssalar$ssalar_paralog_perc_id +
                                             para_comp_within_ssalar$ssalar_paralog_perc_id_r1) / 2
para_comp_within_ssalar_ordered <- para_comp_within_ssalar[order(-para_comp_within_ssalar$average_perc_id), ]
para_comp_within_ssalar_match <- para_comp_within_ssalar_ordered[match(unique(para_comp_within_ssalar_ordered$ensembl_gene_id),
                                       para_comp_within_ssalar_ordered$ensembl_gene_id), ]

## Calculate gene size:
para_comp_within_ssalar_match$gene_1_size <- para_comp_within_ssalar_match$end_position - para_comp_within_ssalar_match$start_position
para_comp_within_ssalar_match$gene_2_size <- para_comp_within_ssalar_match$ssalar_paralog_chrom_end - para_comp_within_ssalar_match$ssalar_paralog_chrom_start

para_comp_within_salmo <- subset(para_comp_within_species,
          ssalar_paralog_subtype == "Salmo")
para_comp_within_salmo$average_perc_id <- (para_comp_within_salmo$ssalar_paralog_perc_id +
                                           para_comp_within_salmo$ssalar_paralog_perc_id_r1) / 2
para_comp_within_salmo_ordered <- para_comp_within_salmo[order(-para_comp_within_salmo$average_perc_id), ]
para_comp_within_salmo_match <- para_comp_within_salmo_ordered[match(unique(para_comp_within_salmo_ordered$ensembl_gene_id),
                                       para_comp_within_salmo_ordered$ensembl_gene_id), ]

## Check if on the same chromosome:
table(para_comp_within_salmo_match$chromosome_name == para_comp_within_salmo_match$ssalar_paralog_chromosome)

## Calculate gene size:
para_comp_within_salmo_match$gene_1_size <- para_comp_within_salmo_match$end_position - para_comp_within_salmo_match$start_position
para_comp_within_salmo_match$gene_2_size <- para_comp_within_salmo_match$ssalar_paralog_chrom_end - para_comp_within_salmo_match$ssalar_paralog_chrom_start

## Calculate difference in gene sizes:
para_comp_within_salmo_match$gene_size <- abs(para_comp_within_salmo_match$gene_1_size - para_comp_within_salmo_match$gene_2_size)

## Plot:
ggplot(data = para_comp_within_salmo_match,
       aes(x = gene_size,
           y = average_perc_id)) +
        geom_point()
```

Subset salmoninae:

```{r, message = FALSE}
para_comp_within_salmoninae <- subset(para_comp_within_species,
          ssalar_paralog_subtype == "Salmoninae")
para_comp_within_salmoninae$average_perc_id <- (para_comp_within_salmoninae$ssalar_paralog_perc_id +
                                                para_comp_within_salmoninae$ssalar_paralog_perc_id_r1) / 2
para_comp_within_salmoninae_ordered <- para_comp_within_salmoninae[order(-para_comp_within_salmoninae$average_perc_id), ]
para_comp_within_salmoninae_match <- para_comp_within_salmoninae_ordered[match(unique(para_comp_within_salmoninae_ordered$ensembl_gene_id),
                                       para_comp_within_salmoninae_ordered$ensembl_gene_id), ]

## Calculate gene size:
para_comp_within_salmoninae_match$gene_1_size <- para_comp_within_salmoninae_match$end_position - para_comp_within_salmoninae_match$start_position
para_comp_within_salmoninae_match$gene_2_size <- para_comp_within_salmoninae_match$ssalar_paralog_chrom_end - para_comp_within_salmoninae_match$ssalar_paralog_chrom_start

## Calculate difference in gene sizes:
para_comp_within_salmoninae_match$gene_size <- abs(para_comp_within_salmoninae_match$gene_1_size - para_comp_within_salmoninae_match$gene_2_size)

## High sequence similarity:
para_comp_within_salmoninae_match_high_sim <- subset(para_comp_within_salmoninae_match,
       average_perc_id >= 85)

## Plot:
ggplot(data = para_comp_within_salmoninae_match,
       aes(x = gene_size,
           y = average_perc_id)) +
        geom_point() +
        geom_point(data = para_comp_within_salmoninae_match_high_sim,
                   aes(x = gene_size,
                       y = average_perc_id),
                   colour = "red")

ggplot(data = para_comp_within_salmoninae_match_high_sim,
       aes(x = gene_size,
           y = average_perc_id)) +
        geom_point()

## Average difference in gene size:
summary(para_comp_within_salmoninae_match_high_sim$gene_size)
```

Compare with pike:

```{r, message = FALSE}
## Identify ensembl gene ids

pike_comp <- getBM(attributes = c('ensembl_gene_id',
                     'chromosome_name',
                     'start_position',
                     'end_position',
                     'elucius_homolog_ensembl_gene',
                     'elucius_homolog_chromosome',
                     'elucius_homolog_chrom_start',
                     'elucius_homolog_chrom_end',
                     'elucius_homolog_orthology_type',
                     'elucius_homolog_perc_id',
                     'elucius_homolog_perc_id_r1',
                     'elucius_homolog_goc_score',
                     #'mmusculus_homolog_wga_coverage',
                     'elucius_homolog_orthology_confidence'), 
      mart = ensembl)

## Subset entries where confidence is '1':
pike_comp_sub <- subset(pike_comp,
       elucius_homolog_orthology_confidence == "1")

## Many to one:
pike_comp_sub_one2many <- subset(pike_comp_sub,
       elucius_homolog_orthology_type == "ortholog_one2many")

pike_comp_sub_one2many$average_perc_id <- (pike_comp_sub_one2many$elucius_homolog_perc_id + pike_comp_sub_one2many$elucius_homolog_perc_id_r1) / 2

## Subset putative immune genes:
pike_comp_sub_one2many_immune <- subset(pike_comp_sub_one2many,
       ensembl_gene_id %in% unique_gene_list &
               average_perc_id >= 75)
               
## Calculate the frequency of pike genes:
pike_comp_sub_one2many_immune_freq <- as.data.frame(table(sort(pike_comp_sub_one2many_immune$elucius_homolog_ensembl_gene)))

## Subset duplicated entries for pike:
pike_comp_sub_one2two_immune <- subset(pike_comp_sub_one2many_immune_freq,
       Freq == 2)$Var1

pike_comp_sub_one2two_immune_list <- as.character(unlist(pike_comp_sub_one2two_immune))

## Subset salmonid genes:
pike_comp_sub_one2many_immune_ohno <- subset(pike_comp_sub_one2many_immune,
       elucius_homolog_ensembl_gene %in% pike_comp_sub_one2two_immune_list)

## Order by pike gene:
pike_comp_sub_one2many_immune_ohno <- pike_comp_sub_one2many_immune_ohno[order(pike_comp_sub_one2many_immune_ohno$elucius_homolog_ensembl_gene), ]

write.table(pike_comp_sub_one2many_immune_ohno,
            file = "results/pike_comp_sub_one2many_immune_ohno_ssalar.txt")

```

Compare with the work of the Bertoletti:

```{r, message = FALSE}
nrow(pike_comp_sub_one2many_immune_ohno)
head(pike_comp_sub_one2many_immune_ohno)
head(salmo_ncbi)

pike_comp_sub_one2many_immune_ohno$ncbi_gene_id <- salmo_ncbi[match(pike_comp_sub_one2many_immune_ohno$ensembl_gene_id,
                 salmo_ncbi$ensembl_gene_id), ]$entrezgene_id

## Read in information from Bertoletti:
bertoletti_data <- read.table(file = "input/supp_table_4_bertolotti.txt",
                              header = TRUE)

bertoletti_data_ss4r <- subset(x = bertoletti_data,
                               type == "ss4r")

## Extract pike matches:
bertoletti_data_ss4r$pike_matches <- gsub(pattern = ":.*",
                                          replacement = "",
                                          bertoletti_data_ss4r$clantree)
bertoletti_data_ss4r$pike_matches <- gsub(pattern = "[(]",
                                          replacement = "",
                                          bertoletti_data_ss4r$pike_matches)
bertoletti_data_ss4r$pike_matches <- gsub(pattern = "_Eluc",
                                          replacement = "",
                                          bertoletti_data_ss4r$pike_matches)

## Convert columns of interest to character variables:
pike_comp_sub_one2many_immune_ohno$ncbi_gene_id <- as.character(pike_comp_sub_one2many_immune_ohno$ncbi_gene_id)
bertoletti_data_ss4r$gene1 <- as.character(bertoletti_data_ss4r$gene1)
bertoletti_data_ss4r$gene2 <- as.character(bertoletti_data_ss4r$gene2)

tmp1 <- row.names(subset(bertoletti_data_ss4r,
       gene1 %in% pike_comp_sub_one2many_immune_ohno$ncbi_gene_id))

tmp2 <- row.names(subset(bertoletti_data_ss4r,
       gene2 %in% pike_comp_sub_one2many_immune_ohno$ncbi_gene_id))

combined <- as.data.frame(table(c(tmp1, tmp2)))

combined <- subset(combined,
                   Freq == 2)

nrow(subset(bertoletti_data_ss4r,
       row.names(bertoletti_data_ss4r) %in% combined$Var1))
```
Match the singletons:

```{r, message = FALSE}



```





```{r, message = FALSE}
pike_comp_sub_one2many_immune_ohno_strutta <- read.table(file = "results/pike_comp_sub_one2many_immune_ohno.txt",
                                                         header = TRUE)

## Convert factors to character lists:
pike_comp_sub_one2many_immune_ohno_strutta$ensembl_gene_id <- as.character(unlist(pike_comp_sub_one2many_immune_ohno_strutta$ensembl_gene_id))

pike_comp_sub_one2many_immune_ohno$ensembl_gene_id <- as.character(unlist(pike_comp_sub_one2many_immune_ohno$ensembl_gene_id))

combined_pike_orts_freq_df <- as.data.frame(table(sort(c(pike_comp_sub_one2many_immune_ohno_strutta$elucius_homolog_ensembl_gene,
                                       pike_comp_sub_one2many_immune_ohno$elucius_homolog_ensembl_gene))))

## Subset pike genes that appear in both comparisons:
combined_pike_orts_freq_ids <- as.character(unlist(subset(combined_pike_orts_freq_df,
          Freq == 4)$Var1))

## Subset salmon matches:
strutta_match <- pike_comp_sub_one2many_immune_ohno_strutta[match(combined_pike_orts_freq_ids,
                                                 pike_comp_sub_one2many_immune_ohno_strutta$elucius_homolog_ensembl_gene), ]

ssalar_match <- pike_comp_sub_one2many_immune_ohno[match(combined_pike_orts_freq_ids,
                                                 pike_comp_sub_one2many_immune_ohno$elucius_homolog_ensembl_gene), ]

head(strutta_match$elucius_homolog_ensembl_gene)
head(ssalar_match$elucius_homolog_ensembl_gene)

## Add one to the other:
ssalar_match$brown_trout_match <- strutta_match$ensembl_gene_id

require(dplyr)
ssalar_to_trutta_matches <- select(salmo_comp_sub,
       "ensembl_gene_id",
       "strutta_homolog_ensembl_gene")

# Subset genes of interest:
tmp <- ssalar_to_trutta_matches[match(ssalar_match$ensembl_gene_id,
                               ssalar_to_trutta_matches$ensembl_gene_id), ]
tmp$combined <- paste(tmp$ensembl_gene_id,
                      "-",
                      tmp$strutta_homolog_ensembl_gene,
                      sep = "")

ssalar_match$combined <- paste(ssalar_match$ensembl_gene_id,
                      "-",
                      ssalar_match$brown_trout_match,
                      sep = "")

table(c(tmp$combined,
        ssalar_match$combined))

```



Check overlap with orthoFinder output:

```{r, message = FALSE}
orthofinder_gene_counts <- read.table(file = "results/putative_strutta_immune_orthogroups_gene_counts.txt",
                                      header = TRUE)

putative_esox_stru_1_to_2 <- subset(orthofinder_gene_counts,
       E_lucius == 1 &
       S_trutta == 2)$Orthogroup

putative_esox_stru_1_to_2 <- as.character(unlist(putative_esox_stru_1_to_2))

## Write to file:
write(x = putative_esox_stru_1_to_2,
      file = "results/putative_esox_stru_1_to_2_orthogroups.txt")

## Read in associated gene ids:
putative_esox_stru_1_to_2_ids <- read.table(file = "results/putative_esox_stru_1_to_2_orthogroups_gene_ids.txt",
                                            header = FALSE)

## Check overlap with Ensembl analysis:
putative_esox_stru_1_to_2_ids$V5 <- putative_esox_stru_1_to_2_ids$V2
putative_esox_stru_1_to_2_ids[] <- lapply(putative_esox_stru_1_to_2_ids,
                                          as.character)

## Split and recombine:
tmp_1 <- putative_esox_stru_1_to_2_ids[, c(1:3)]
tmp_2 <- putative_esox_stru_1_to_2_ids[, c(1,4:5)]
colnames(tmp_1) <- c("V1", "V4", "V5")

tmp <- rbind(tmp_1,
             tmp_2)

## Remove extensions and trailing characters:
tmp$V4 <- gsub(pattern = "[.]1",
     replacement = "",
     tmp$V4)

tmp$V5 <- gsub(pattern = "[.]1,",
               replacement = "",
               tmp$V5)

tmp$V5 <- gsub(pattern = "[.]1",
               replacement = "",
               tmp$V5)

## Subset putative ohnolog gene names:
require(dplyr)
pike_comp_sub_one2many_immune_ohno_ids <- select(pike_comp_sub_one2many_immune_ohno,
       "ensembl_gene_id",
       "elucius_homolog_ensembl_gene")

subset(tmp, V1 == "OG0008292")

tmp_3 <- tmp[na.omit(match(tmp$V5,
                pike_comp_sub_one2many_immune_ohno_ids$ensembl_gene_id)), ]

```
